src/Codata/Guarded/Stream.agda

```diff
@@ -34,118 +37,144 @@ record Stream (A : Set a) : Set a where
-head (iterate f x) = x
-tail (iterate f x) = iterate f (f x)
-
-unfold : (A → A × B) → A → Stream B
-head (unfold next seed) = P.proj₂ (next seed)
-tail (unfold next seed) = unfold next (P.proj₁ (next seed))
+iterate f = unfold < f , id >
 
 nats : Stream ℕ
-nats = iterate suc zero
+nats = tabulate id
 
-tabulate : (ℕ → A) → Stream A
-head (tabulate f) = f zero
-tail (tabulate f) = tabulate (f ∘ suc)
+------------------------------------------------------------------------
+-- Lookup
+
+lookup : ℕ → Stream A → A
+lookup zero = head
+lookup (suc n) = lookup n ∘′ tail
+
+_[_] : Stream A → ℕ → A
+_[_] = flip lookup
+
+------------------------------------------------------------------------
+-- Transforming streams
 
 map : (A → B) → Stream A → Stream B
-head (map f xs) = f (head xs)
-tail (map f xs) = map f (tail xs)
+map f s .head = f (s .head)
+map f s .tail = map f (s .tail)
 
 ap : Stream (A → B) → Stream A → Stream B
-head (ap fs xs) = head fs (head xs)
-tail (ap fs xs) = ap (tail fs) (tail xs)
+ap fs xs .head = fs .head (xs .head)
+ap fs xs .tail = ap (fs .tail) (xs .tail)
+
+scanl : (B → A → B) → B → Stream A → Stream B
+scanl c n s .head = n
+scanl c n s .tail = scanl c (c n (s .head)) (s .tail)
 
 zipWith : (A → B → C) → Stream A → Stream B → Stream C
-head (zipWith f xs ys) = f (head xs) (head ys)
-tail (zipWith f xs ys) = zipWith f (tail xs) (tail ys)
+zipWith f s t .head = f (s .head) (t .head)
+zipWith f s t .tail = zipWith f (s .tail) (t .tail)
 
-scanl : (B → A → B) → B → Stream A → Stream B
-head (scanl f b xs) = b
-tail (scanl f b xs) = scanl f (f b (head xs)) (tail xs)
+transpose : List (Stream A) → Stream (List A)
+transpose [] = repeat []
+transpose (s ∷ ss) = zipWith _∷_ s (transpose ss)
 
 tails : Stream A → Stream (Stream A)
-head (tails xs) = xs
-tail (tails xs) = tails (tail xs)
+tails s .head = s
+tails s .tail = tails (s .tail)
 
-splitAt : (n : ℕ) → Stream A → Vec A n × Stream A
-splitAt 0 xs = [] , xs
-splitAt (suc n) xs = P.map₁ (head xs ∷_) (splitAt n (tail xs))
+evens : Stream A → Stream A
+evens s .head = s .head
+evens s .tail = evens (s .tail .tail)
 
-take : (n : ℕ) → Stream A → Vec A n
-take n xs = P.proj₁ (splitAt n xs)
+odds : Stream A → Stream A
+odds s = evens (s .tail)
 
-drop : (n : ℕ) → Stream A → Stream A
-drop n xs = P.proj₂ (splitAt n xs)
+------------------------------------------------------------------------
+-- List⁺-related functions
 
-chunksOf : (n : ℕ) → Stream A → Stream (Vec A n)
-head (chunksOf n xs) = take n xs
-tail (chunksOf n xs) = chunksOf n (drop n xs)
+infixr 5 _⁺++_
+_⁺++_ : List⁺ A → Stream A → Stream A
+(x ∷ xs) ⁺++ ys = x ∷ xs ++ ys
 
-interleave : Stream A → Stream A → Stream A
-head (interleave xs ys) = head xs
-tail (interleave xs ys) = interleave ys (tail xs)
+concat : Stream (List⁺ A) → Stream A
+concat {A = A} = λ s → [] ++concat s -- Here using λ so that _++concat_ doesn't take an extra argument
+  module Concat where
+    _++concat_ : List A → Stream (List⁺ A) → Stream A
```

Comment:
Why define an infix name if it's to use it suffix?

Reply:
How about you call it just `++-concat` and then call it using `++-concat []` on Line 117?